src/main/java/org/myrobotlab/framework/Plan.java

```diff
@@ -1,34 +1,209 @@
 package org.myrobotlab.framework;
 
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
 import java.util.Map;
+import java.util.Set;
 import java.util.TreeMap;
 
-/**
- * 
- * A convenient wrapper to show what peers and services "could" be created
- * depending on current MetaData and ServiceData overrides
- *
- */
+import org.myrobotlab.logging.LoggerFactory;
+import org.myrobotlab.service.config.ArduinoConfig;
+import org.myrobotlab.service.config.OpenCVConfig;
+import org.myrobotlab.service.config.SerialConfig;
+import org.myrobotlab.service.config.ServiceConfig;
+import org.myrobotlab.service.meta.abstracts.MetaData;
+import org.slf4j.Logger;
+
 public class Plan {
   /**
-   * Map<{actualName}, {Type}>
+   * the root of the tree of configuration
    */
-  Map<String, String> services = new TreeMap<>();
+  final String name;
+
+  public final static Logger log = LoggerFactory.getLogger(Plan.class);
+  
+  LinkedHashMap<String, ServiceConfig> config = new LinkedHashMap<>();
+  
+  public Map<String, Map<String, ServiceReservation>> peers = new TreeMap<String, Map<String, ServiceReservation>>();
+
+  // final MetaData metaData;
+
+  public Plan(String name) {
+    this.name = name;
+  }
 
   public String toString() {
     StringBuilder sb = new StringBuilder("\n");
-    for (Map.Entry<String, String> e : services.entrySet()) {
+    for (Map.Entry<String, ServiceConfig> e : config.entrySet()) {
       sb.append(e.getKey()).append("=").append(e.getValue()).append("\n");
     }
     return sb.toString();
   }
 
-  public void put(String serviceName, String serviceType) {
-    services.put(serviceName, serviceType);
+  public ServiceConfig put(String name, ServiceConfig sc) {
+    return config.put(name, sc);
+  }
+
+  public void putAll(Map<String, ServiceConfig> c) {
+    config.putAll(c);
+  }
+
+  public ServiceConfig get(String name) {
+    return config.get(name);
+  }
+
+  public Set<String> keySet() {
+    // return a copy of the keyset to avoid concurrent modification
+    Set<String> ret = new LinkedHashSet<>(config.keySet());
```

Comment:
why the type conversion here to LinkedHashSet?  if config.KeySet() isn't ordered, this is pointless.

Reply:
config.keySet() returns a LinkedKeySet ... because config is a LinkedHashMap..  this type casting is probably not necessary.

Reply:
Yeah ... its muscle memory for me ..
your always supposed to return a generalized interface